File: rails/wx-services/components/wunder_conditions.rb
require 'logger'
require 'date'
require 'net/http'
require 'open-uri'
require 'rexml/document'
include REXML

# http://www.weather.gov/xml/current_obs/KBVY.xml

#mysql> describe noaa_conditions;
#+----------------+-------------+------+-----+---------+----------------+
#| Field          | Type        | Null | Key | Default | Extra          |
#+----------------+-------------+------+-----+---------+----------------+
#| id             | int(11)     | NO   | PRI | NULL    | auto_increment |
#| created_at     | datetime    | YES  | MUL | NULL    |                |
#| location       | varchar(30) | NO   | MUL |         |                |
#| updated_at     | datetime    | YES  |     | NULL    |                |
#| conditions     | text        | NO   |     | NULL    |                |
#| as_of          | datetime    | YES  | MUL | NULL    |                |
#| visibility     | bigint(20)  | YES  |     | NULL    |                |
#| conditions_xml | text        | YES  |     | NULL    |                |
#+----------------+-------------+------+-----+---------+----------------+
#8 rows in set (0.01 sec)

class WunderConditionsWriter
  log = Logger.new(STDOUT)
  log.level = Logger::INFO

  URL='http://api.wunderground.com/auto/wui/geo/WXCurrentObXML/index.xml?query=' + AppConfig.noaa_location
  f = open(URL) # open-uri - treat the URL as an input stream
  data = f.read
  log.debug(data)
  doc = Document.new(data)
  location = doc.elements[1].elements["station_id"].text
  log.debug("location: " + location)
  as_of = Time.rfc822(doc.elements[1].elements["observation_time_rfc822"].text).utc
  log.debug("as_of: " + as_of.to_s)

  visibility = doc.elements[1].elements["visibility_mi"].text.to_i
  log.debug("visibility: " + visibility.to_s)

  conditions = doc.elements[1].elements["weather"].text
  log.debug("conditions: " + conditions)

  record = WunderConditions.find_or_create_by_location(location)
  log.debug(record)
  record[:conditions_xml] = data
  record[:location] = location
  record[:as_of] = as_of
  record[:conditions] = conditions
  record[:visibility] = visibility
  record[:updated_at] = Time.now
  log.debug(record)
  record.save!
end

File: gae/weather/services/wunder_conditions.py
import datetime
import pytz
from google.appengine.api import memcache
from xml.etree import ElementTree
import urllib

#http://api.wunderground.com/auto/wui/geo/WXCurrentObXML/index.xml?query=kbvy

URL='http://api.wunderground.com/auto/wui/geo/WXCurrentObXML/index.xml?query='

class WunderConditions:
  def __init__(self, as_of, weather, visibility):
    self.as_of = as_of
    self.weather = weather
    self.visibility = visibility
    self.retrieved = datetime.datetime.utcnow().replace(tzinfo=pytz.utc)

class ConditionsFactory:
  @staticmethod
  def get(location):
    return memcache.get(location, namespace='conditions')

  @staticmethod
  def put(location):
    feed = urllib.urlopen(URL+location)
    tree = ElementTree.parse(feed)
    as_of = tree.find('/observation_time').text
    weather = tree.find('/weather').text
    visibility = tree.find('/visibility_mi').text
    if as_of:
      conditions = WunderConditions(as_of, weather, visibility)
      memcache.set(location, conditions, time=7200, namespace='conditions')
      return conditions
